README.md

```diff
@@ -438,22 +446,78 @@ _To be added_
 
 ### 3. Other configuration options
 
-SGX-LKL-OE has a number of other configuration options e.g. for configuring the
-in-enclave scheduling, network configuration, or debugging/tracing. To see all
-options, run:
+SGX-LKL-OE has a number of other configuration options, e.g. for configuring the
+disks, in-enclave scheduling, network configuration, or debugging/tracing.
+
+There are two ways to specify configuration options: environment variables and JSON files.
+
+During debugging, it can be convenient to specify debugging options as environment
+variables. Note that for debugging options to have an effect, SGX-LKL must be built
+with `DEBUG=true`.
+To see a list of all environment configuration variables, run:
 ```
-sgx-lkl-run-oe --help
+sgx-lkl-run-oe --help-config
 ```
 
-Note that for the debugging options to have an effect, SGX-LKL must be built
-with `DEBUG=true`.
+The recommended way to configure SGX-LKL is through host config and app config JSON files.
+Using JSON files is also required when creating Confidential Containers (see next section).
+To create initial host and app config files in the current directory, run:
+```
+sgx-lkl-cfg create --disk sgxlkl-disk.img
+```
+Note that you may have to adapt the generated files to your needs.
+See the `tests/containers/cc` folder for example host and app config files.
+You can then run SGX-LKL using the host and app config files alone:
+```
+sgx-lkl-run-oe --hw-debug --host-config=host-config.json --app-config=app-config.json
+```
+Note that environment variables can still be used to specify additional options.
+This is especially useful for host configuration options that are only known
+at runtime in the target system where SGX-LKL will run, for example the name
+of the network tap device (`SGXLKL_TAP`).
+
+F. Confidential Container bundling for deployment
+-------------------------------------------------
+
+A Confidential Container is a Docker image that only contains the disk images
+and the host and app config files.
+It is used for deploying an SGX-LKL application to a target system that has
+Docker as well as SGX-LKL installed on the host.
+This allows to update SGX-LKL without rebuilding all Docker images.
+Like Docker, SGX-LKL can be considered part of the host runtime environment.
+
+To create a Confidential Container, run:
+```
+sgx-lkl-docker build-cc --name=my-app-cc --mode=debug --host-cfg=host-config.json --app-cfg=app-config.json
```

Comment:
I wonder if it is right that we have to specify the execution mode at CC build time. Why is that necessary?

Reply:
This determines the ENTRYPOINT that's used. E.g. `--mode=debug` results in `ENTRYPOINT ["/opt/sgx-lkl-debug/bin/sgx-lkl-run-oe", "--host-config=/host-cfg.json", "--app-config=/app-cfg.json"]`. In principle you can also leave out the entrypoint and specify it at runtime but that gets a bit unwieldy. Unfortunately it's not possible to do `-v /opt/sgx-lkl-debug:/opt/sgx-lkl` because the installation prefix must stay the same as it's baked in to the executable's interpreter path which must be absolute (`/opt/sgx-lkl-debug/lib/external/ld-linux-x86-64.so.2`). One thing we could simplify is to change the install prefixes to `/opt/sgx-lkl/<build-type>` so that at least the mount flag is always the same regardless of build type: `-v /opt/sgx-lkl:/opt/sgx-lkl`.

Reply:
> This determines the ENTRYPOINT that's used. E.g. `--mode=debug` results in `ENTRYPOINT ["/opt/sgx-lkl-debug/bin/sgx-lkl-run-oe", "--host-config=/host-cfg.json", "--app-config=/app-cfg.json"]`. In principle you can also leave out the entrypoint and specify it at runtime but that gets a bit unwieldy. Unfortunately it's not possible to do `-v /opt/sgx-lkl-debug:/opt/sgx-lkl` because the installation prefix must stay the same as it's baked in to the executable's interpreter path which must be absolute (`/opt/sgx-lkl-debug/lib/external/ld-linux-x86-64.so.2`). One thing we could simplify is to change the install prefixes to `/opt/sgx-lkl/<build-type>` so that at least the mount flag is always the same regardless of build type: `-v /opt/sgx-lkl:/opt/sgx-lkl`.

Hmm, I don't think that the above is a great justification to have the `--mode` flag. You could argue that an entrypoint that simply refers to the SGX-LKL launcher is something that should come from the host environment and not be part of the CC. What would it take to provide the entrypoint at launch time?

Reply:
```
docker run --rm --privileged --network=host \
  -v /opt/sgx-lkl-debug:/opt/sgx-lkl-debug \
  -v /etc/ssl/certs:/etc/ssl/certs:ro \
  --entrypoint=/opt/sgx-lkl-debug/bin/sgx-lkl-run-oe
  my-app-cc \
  --host-config=/host-cfg.json --app-config=/app-cfg.json \
  --hw-debug
```
We could get rid of `--host-config` and `--app-config` if we extend SGX-LKL to pick up the paths from environment variables, then we can specify them in the CC Dockerfile via `ENV SGXLKL_APP_CONFIG_PATH=/app-cfg.json` etc.

Reply:
Note that I agree with you, it would be better to get rid of the `--mode` flag. If we need to add it in the future for other purposes again we can bring it back.

Reply:
Alternatively, instead of new environment variables, we could have SGX-LKL look for default paths in the current directory, like `host-config.json` and `app-config.json`.
The default ENTRYPOINT would be `/opt/sgx-lkl-release/bin/sgx-lkl-run-oe` and the default CMD would be `--hw-release`. Then the Docker run would be:
```
docker run --rm --privileged --network=host \
  -v /opt/sgx-lkl-release:/opt/sgx-lkl-release \
  -v /etc/ssl/certs:/etc/ssl/certs:ro \
  my-app-cc
```

If you want to use something non-default, you can specify the entrypoint and SGX-LKL arguments, e.g here for running a debug build in sw mode:
```
docker run --rm --privileged --network=host \
  -v /opt/sgx-lkl-debug:/opt/sgx-lkl-debug \
  -v /etc/ssl/certs:/etc/ssl/certs:ro \
  --entrypoint=/opt/sgx-lkl-debug/bin/sgx-lkl-run-oe \
  my-app-cc --run-sw
```

Reply:
Sorry for the spam, but I'm a bit torn here. The only reason for running a CC in a different mode is for debugging/testing, and in that case you would very likely not use the same CC anyway as you would not want attestation-backed disk encryption, and maybe you also compile your app differently if its a native app (e.g. with debug symbols etc.). What `--mode` then does is two things: first, it makes running the CC easier by hard-coding some things, and second it could check for common mistakes, e.g. warning if the disks are not encrypted when using `--mode=release`. Are you sure we want to get rid of that?